daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DFDLXTypeCalcFunctions.scala

```diff
@@ -26,15 +26,17 @@ import org.apache.daffodil.dpath.NodeInfo.PrimType.IntegerKind
 import org.apache.daffodil.dpath.NodeInfo.PrimType.String
 import org.apache.daffodil.dpath.NodeInfo.PrimType.Integer
 import org.apache.daffodil.processors.TypeCalculator
+import org.apache.daffodil.infoset.DataValue.DataValuePrimitive
+import org.apache.daffodil.infoset.DataValue.DataValuePrimitive
 
 case class DFDLXInputTypeCalc(typedRecipes: List[(CompiledDPath, NodeInfo.Kind)], calc: TypeCalculator[AnyRef, AnyRef])
   extends FNTwoArgs(typedRecipes.map(_._1)) {
 
   val srcType = typedRecipes(1)._2
 
-  override def computeValue(arg1: AnyRef, arg2: AnyRef, dstate: DState): AnyRef = {
-    calc.inputTypeCalcRun(dstate, arg2, srcType)
-    dstate.currentValue
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    calc.inputTypeCalcRun(dstate, arg2.getAnyRef, srcType)
+    dstate.currentValue.getNonNullable
```

Comment:
I don't think this will actually fail if currentValue is NoValue.

Not that we were checking for that case before this refactor, but maybe an assertion here would be warrented?